OpenRA.Mods.Cnc/Traits/SupportPowers/GrantPrerequisiteChargeDrainPower.cs

```diff
@@ -0,0 +1,211 @@
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[Desc("Grants a prerequisite while discharging at a configurable rate.")]
+	public class GrantPrerequisiteChargeDrainPowerInfo : SupportPowerInfo, ITechTreePrerequisiteInfo
+	{
+		[Desc("Rate at which the power discharges compared to charging")]
+		public readonly int DischargeModifier = 300;
+
+		[FieldLoader.Require]
+		[Desc("The prerequisite type that this provides.")]
+		public readonly string Prerequisite = null;
+
+		[Translate]
+		[Desc("Label to display over the support power icon and in its tooltip while the power is active.")]
+		public readonly string ActiveText = "ACTIVE";
+
+		[Translate]
+		[Desc("Label to display over the support power icon and in its tooltip while the power is available but not active.")]
+		public readonly string AvailableText = "READY";
+
+		IEnumerable<string> ITechTreePrerequisiteInfo.Prerequisites(ActorInfo info)
+		{
+			yield return Prerequisite;
+		}
+
+		public override object Create(ActorInitializer init) { return new GrantPrerequisiteChargeDrainPower(init.Self, this); }
+	}
+
+	public class GrantPrerequisiteChargeDrainPower : SupportPower, ITechTreePrerequisite, INotifyOwnerChanged
+	{
+		readonly GrantPrerequisiteChargeDrainPowerInfo info;
+		TechTree techTree;
+		bool active;
+
+		public GrantPrerequisiteChargeDrainPower(Actor self, GrantPrerequisiteChargeDrainPowerInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+		}
+
+		protected override void Created(Actor self)
+		{
+			// Special case handling is required for the Player actor.
+			// Created is called before Player.PlayerActor is assigned,
+			// so we must query other player traits from self, knowing that
+			// it refers to the same actor as self.Owner.PlayerActor
+			var playerActor = self.Info.Name == "player" ? self : self.Owner.PlayerActor;
+
+			techTree = playerActor.Trait<TechTree>();
+
+			base.Created(self);
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			techTree = newOwner.PlayerActor.Trait<TechTree>();
+			active = false;
+		}
+
+		public override SupportPowerInstance CreateInstance(string key, SupportPowerManager manager)
+		{
+			return new DischargableSupportPowerInstance(key, info, manager);
+		}
+
+		public void Activate(Actor self, SupportPowerInstance instance)
+		{
+			active = true;
+			techTree.ActorChanged(self);
+		}
+
+		public void Deactivate(Actor self, SupportPowerInstance instance)
+		{
+			active = false;
+			techTree.ActorChanged(self);
+		}
+
+		IEnumerable<string> ITechTreePrerequisite.ProvidesPrerequisites
+		{
+			get
+			{
+				if (!active)
+					yield break;
+
+				yield return info.Prerequisite;
+			}
+		}
+
+		public class DischargableSupportPowerInstance : SupportPowerInstance
```

Comment:
Is it convention to put `*SupportPowerInstance`s in the same file as the `*Power`?
This may hurt discoverability of this type (if IDE tooling is busted for some reason [Omnisharp isn't too reliable at times]).

Reply:
This is an internal implementation detail that is only useful for things that are intentionally messing with the internal behaviour of the support power (the `DrainPrerequisitePowerOnDamage` trait), so yes.

Reply:
Spelling? Dischargeable(?) not sure about this tho..